subprojects/model-core/src/integTest/groovy/org/gradle/api/provider/PropertyAssignmentIntegrationTest.groovy

```diff
@@ -0,0 +1,405 @@
+        "Collection<T> += Iterable<T>"           | "+="      | "MutableList<String>"        | "ListProperty<String>"        | 'listOf("a") as Iterable<String>'              | 'Result: [a]'                                 | failsWithDescription("Unresolved reference. None of the following candidates is applicable because of receiver type mismatch:")
+        "Collection<T> += Provider<Iterable<T>>" | "+="      | "MutableList<String>"        | "ListProperty<String>"        | 'provider { listOf("a") as Iterable<String> }' | failsWithDescription("Type mismatch")         | failsWithDescription("Unresolved reference. None of the following candidates is applicable because of receiver type mismatch:")
+        "Map<K, V> = Map<K, V>"                  | "="       | "Map<String, String>"        | "MapProperty<String, String>" | 'mapOf("a" to "b")'                            | 'Result: {a=b}'                               | 'Result: {a=b}'
+        "Map<K, V> = Provider<Map<K, V>>"        | "="       | "Map<String, String>"        | "MapProperty<String, String>" | 'provider { mapOf("a" to "b") }'               | failsWithDescription("Type mismatch")         | 'Result: {a=b}'
+        "Map<K, V> += Pair<K, V>"                | "+="      | "MutableMap<String, String>" | "MapProperty<String, String>" | '"a" to "b"'                                   | 'Result: {a=b}'                               | failsWithDescription("Unresolved reference. None of the following candidates is applicable because of receiver type mismatch:")
+        "Map<K, V> += Provider<Pair<K, V>>"      | "+="      | "MutableMap<String, String>" | "MapProperty<String, String>" | 'provider { "a" to "b" }'                      | failsWithDescription("None of the following") | failsWithDescription("Unresolved reference. None of the following candidates is applicable because of receiver type mismatch:")
+        "Map<K, V> += Map<K, V>"                 | "+="      | "MutableMap<String, String>" | "MapProperty<String, String>" | 'mapOf("a" to "b")'                            | 'Result: {a=b}'                               | failsWithDescription("Unresolved reference. None of the following candidates is applicable because of receiver type mismatch:")
+        "Map<K, V> += Provider<Map<K, V>>"       | "+="      | "MutableMap<String, String>" | "MapProperty<String, String>" | 'provider { mapOf("a" to "b") }'               | failsWithDescription("None of the following") | failsWithDescription("Unresolved reference. None of the following candidates is applicable because of receiver type mismatch:")
+    }
+
+    def "test Groovy FileCollection types assignment for #description"() {
```

Comment:
It would be more appropriate to move FileCollection tests in file-collections project

Reply:
I left it in the same file for now.